### ISSUE TITLE
Renew token

### ISSUE THREAD

**Issue by simplenotezy** [Tmeister/wp-api-jwt-auth#186]
I get errors “Token is Expired”. How can we renew/extend it?

**Comment by pesseba** [comment]
you can use the hook bellow:

```
function on_jwt_expire_token($exp){
	$days = 182;
	$exp = time() + (86400 * $days);		
	return $exp;
}
add_filter('jwt_auth_expire', 'on_jwt_expire_token' ,10,2);
```

**Comment by simplenotezy** [comment]
No, I mean refresh it. Not extend expiry time

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by pesseba** [comment]
The JWT class inside the plugin has a global variable called $leeway. You must set a value to subtract time from timestamp and to provide some extra leeway time to account for clock skew.
`JWT::$leeway = 123; //datetime in number `

**Comment by simplenotezy** [comment]
@pesseba i want the token to be valid for the default time period. Then I want to occasionally send the token to some kind of refresh endpoint and receive a fresh token that's valid for the default time period. Makes sense?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by pesseba** [comment]
@simplenotezy I believe you want a fresh token and deprecate the older one. I don't know how to deprecate a token (if possible)... But, if you want an endpoint to only refresh the token you can do this:

```
public function request_refresh_token($request){
		
	$r = new WP_REST_Request( "POST", "/jwt-auth/v1/token/validate");
	$token = $request->get_header("Authorization");
	$r->set_header("Authorization",$token);
	$response = rest_do_request( $r );	
	
	if(isset($response->data["code"]) && $response->data["code"] == "jwt_auth_valid_token"){
		
		$user = wp_get_current_user();
		add_filter( "authenticate", function($name,$pwd){ return $user; },10,2);
	
		$r = new WP_REST_Request( "POST", "/jwt-auth/v1/token");
		$response = rest_do_request( $r );
		return new WP_REST_Response($response);		
	}
	
	return new WP_REST_Response($response->data);
}

function on_rest_api_init(){

	register_rest_route("jwt-auth/v1", "refresh", array(
		array(
				"methods" => "GET",
				"callback" => "request_refresh_token",					
			),				
		)
	);		
}
add_action("rest_api_init","on_rest_api_init",10);
```

**Comment by dvkch** [comment]
It would be amazing if a call to generate a new token on condition that the existing one is valid was supported. It could be disabled by default to warn people that it doesn't invalidate the previous one. 

The code sample above works with some tweaks but feels very hacky as it needs to hook wordpress authentication methods and calls into this plugins REST methods, while adapting a bit the plugin to put the token generation code in a helper method and reusing it in two different endpoint would be a lot clearer.

In the mean time if anyone wants to try this here is the code I used, adapted from @pesseba  example above : 

```
// Temporary authentication bypass to refresh JWT token
function jwt_bypass_authenticate($user, $username, $password) {
    $user = get_user_by("login", $username);
    if (!is_null($user) && $password === "jwt_refresh" . $user->user_pass) {
        return $user;
    }
    return null;
}

// REST endpoint to refresh JWT token given a valid input token, inspired by
// https://github.com/Tmeister/wp-api-jwt-auth/issues/186#issuecomment-597786797
function request_refresh_token($request) {
    // ensures token is valid
    $r = new WP_REST_Request("POST", "/jwt-auth/v1/token/validate");
    $token = $request->get_header("Authorization");
    $r->set_header("Authorization",$token);
    $response = rest_do_request( $r );
    
    if (isset($response->data["code"]) && $response->data["code"] == "jwt_auth_valid_token") {
        // replay login
        $user = wp_get_current_user();
        $r = new WP_REST_Request( "POST", "/jwt-auth/v1/token");
        $r->set_param("username", $user->user_login);
        $r->set_param("password", "jwt_refresh" . $user->user_pass);

        // bypass authentication
        add_filter("authenticate", "jwt_bypass_authenticate", 10, 3);
        $response = rest_do_request( $r );
        remove_filter("authenticate", "jwt_bypass_authenticate", 10);

        // return new token
        return new WP_REST_Response($response->data);
    }
    
    // forward validation error
    return new WP_REST_Response($response->data);
}

function add_request_refresh_token() {
    register_rest_route("jwt-auth/v1/token", "refresh", array(
        "methods" => "POST", "callback" => "request_refresh_token"
    ));
}
add_action("rest_api_init", "add_request_refresh_token", 10);
```

**Comment by pesseba** [comment]
The add_filter("authenticate",... is just needed to bypass username and password. The filter only operates in this specific endpoint call... So, don't worry.

**Comment by dvkch** [comment]
For sure, but the process as a whole still depends on the plugins interface, while not being tested alongside it for possible breaking changes, even though it's the public API and shouldn't change. Since it seems to be sometimes requested here and on wordpress.org you see why I was wondering if this could be supported officially. But I totally understand if this is not a priority or not in the plugin scope!

**Comment by gassan** [comment]
https://github.com/Tmeister/wp-api-jwt-auth/pull/115
merged to v1.27
